packages/datadog-plugin-fastify/src/index.js

```diff
@@ -0,0 +1,104 @@
+'use strict'
+
+const web = require('../../dd-trace/src/plugins/util/web')
+
+function createWrapFastify (tracer, config) {
+  config = web.normalizeConfig(config)
+
+  return function wrapFastify (fastify) {
+    const fastifyWithTrace = function fastifyWithTrace () {
+      return fastify._datadog_wrapper.apply(this, arguments)
+    }
+
+    fastify._datadog_wrapper = function () {
+      const app = fastify.apply(this, arguments)
+
+      app.addHook('onRequest', createOnRequest(tracer, config))
+      app.addHook('preHandler', preHandler)
```

Comment:
Why do we hook into the `preHandler` and not the `onSend` or `onError`.

Also, can we also wrap `send` using the `reply` in `onRequest`?

Reply:
> Why do we hook into the preHandler and not the onSend or onError.

This is to make sure we wrap before the handler, and `preHandler` is the hook that happens before the handler and receives a Fastify request in the most consistent way. `onError` doesn't exist in older versions, and `onSend` happens to late in the lifecycle.

> Also, can we also wrap send using the reply in onRequest?

In older versions, `onRequest` didn't pass `request` and instead passed the raw request.